lib/nmatrix/math.rb

```diff
@@ -364,6 +364,37 @@ def cov(opts={})
     deviation_scores.transpose.dot(deviation_scores) / denominator
   end
 
+  # Raise a square matrix to a power. Be careful of buffer overflows!
+  # In case *n* is 0, a matrix of ones of the same dimension is returned. In case
+  # of negative *n*, the matrix is inverted and the absolute value of *n* taken 
+  # for computing the power.
+  # 
+  # == Arguments
+  # 
+  # * +n+ - Integer to which self is to be raised.
+  def pow n
+    raise ShapeError, "Only works with 2D square matrices." if shape[0] != shape[1]
+    raise TypeError, "Only works with integer powers" unless n.is_a?(Integer)
+
+    product = NMatrix.diag [1]*shape[0] 
+    sequence = self.clone
+
+    if n == 0
+      return NMatrix.ones(shape)
+    elsif n < 0
+      n = n.abs
+      sequence.invert!
+    end
+
+    while n > 0
+      product = product.dot(sequence) if n % 2 == 1
+      n = n / 2
+      sequence = sequence.dot(sequence)
```

Comment:
The `sequence = sequence.dot(sequence)` part is superfluous when n = 1

In general, this while loop doesn't feel very ruby-ish.  (It took me rather a while to figure out that what you're doing here is decomposing n into powers of two and then doing straight multiplication for what's left over.)  I probably would have had a separate function to do the decomposition of n, and then compute the appropriate powers.  This is probably marginally more efficient as you have it, though.  At minimum, can you put a comment here pointing out what this loop is doing?